services/repository/commitstatus/commitstatus.go

```diff
@@ -26,12 +27,41 @@ func getCacheKey(repoID int64, brancheName string) string {
 	return fmt.Sprintf("commit_status:%x", hashBytes)
 }
 
-func updateCommitStatusCache(ctx context.Context, repoID int64, branchName string, status api.CommitStatusState) error {
+type commitStatusCacheValue struct {
+	State     string `json:"state"`
+	TargetURL string `json:"target_url"`
+}
+
+func getCommitStatusCache(repoID int64, branchName string) *commitStatusCacheValue {
+	c := cache.GetCache()
+	statusStr, ok := c.Get(getCacheKey(repoID, branchName)).(string)
+	if ok && statusStr != "" {
+		var cv commitStatusCacheValue
+		err := json.Unmarshal([]byte(statusStr), &cv)
```

Comment:
That's a mess .... why not wait for #30417 first? Could you help to resolve the conflicts with #30417 ?

Reply:
Because this PR will be backport. I will contribute to #30417 and resolve the conflicts.

Reply:
#30417 also must be backported because it fixes various cache bugs.

Reply:
The original description explained why it is necessary and how the bugs are fixed, but the content was edited.

![image](https://github.com/go-gitea/gitea/assets/2114189/0d550a37-c51c-4e79-ac83-34d1c832cfb3)

Reply:
This PR also needs to be backpor to v1.21.